server/src/main/java/uk/ac/cam/echo/extrusionfinder/preprocessor/ProfileDetector.java

```diff
@@ -0,0 +1,318 @@
+        Mat imageResized = new Mat();
+        Mat output = new Mat();
+        Size newSize = new Size(diameter, diameter);
+        Imgproc.resize(input, imageResized, newSize);
+
+        // Assuming RGB image already.
+        output = imageResized;
+
+        return output;
+    }
+
+    /**
+     * Applies a bilateral blur to the given image.
+     * <p>
+     * The input image is blurred using supplied parameters.
+     * <p>
+     * See http://docs.opencv.org/modules/imgproc/doc/filtering.html#bilateralfilter and
+     * http://www.dai.ed.ac.uk/CVonline/LOCAL_COPIES/MANDUCHI1/Bilateral_Filtering.html
+     */
+    private Mat blur(Mat input, int blurDiameter, int sigmaColor, int sigmaSpace) {
+        Mat output = new Mat();
+        Imgproc.bilateralFilter(input, output, blurDiameter, sigmaColor, sigmaSpace);
+        return output;
+    }
+
+    /**
+     * Converts the input image to greyscale using the minimum component per pixel.
+     * <p>
+     * For each pixel, the greyscale value is given as min(pixel.red, pixel.green, pixel.blue).
+     */
+    private Mat minGreyscale(Mat input) {
+        // Extract individual channels.
+        Mat c1 = new Mat(diameter, diameter, CvType.CV_8UC1);
+        Mat c2 = new Mat(diameter, diameter, CvType.CV_8UC1);
+        Mat c3 = new Mat(diameter, diameter, CvType.CV_8UC1);
+        List<Mat> matSrc = new ArrayList<Mat>(3);
+        List<Mat> matDst = new ArrayList<Mat>(3);
+        MatOfInt fromTo = new MatOfInt(0, 0, 1, 1, 2, 2);
+        matSrc.add(input);
+        matDst.add(c1);
+        matDst.add(c2);
+        matDst.add(c3);
+        Core.mixChannels(matSrc, matDst, fromTo);
+
+        // Get minimums
+        Mat output;
+        output = c1;
+        Core.min(output, c2, output);
+        Core.min(output, c3, output);
+
+        return output;
+    }
+
+    /**
+     * Converts the input image to greyscale using the averaging of all components per pixel.
+     * <p>
+     * For each pixel, the greyscale value is given as (pixel.red + pixel.green + pixel.blue) / 3.
+     */
+    private Mat avgGreyscale(Mat input) {
+        Mat output = new Mat();
+        Imgproc.cvtColor(input, output, Imgproc.COLOR_RGB2GRAY);
+        return output;
+    }
+
+    // This function needs to be revised!
+    /**
+     * Applies a radial fade to black to the given image from the center.
+     * <p>
+     * The fade is linear, and the fading region is the circle, centred with the image, with double
+     * the diameter.
+     */
+    private Mat fade(Mat input) {
+        byte[] bytes = new byte[diameter * diameter];
+        Mat mask = new Mat(diameter, diameter, CvType.CV_8UC1);
+
+        int r = diameter / 2;
+        int i = 0;
+        for (int y = 0; y < diameter; y++) {
+            for (int x = 0; x < diameter; x++, i++) {
+                // This could be normalised!
+                double keep = 0.707106781187 - (Math.sqrt((x-r)*(x-r) + (y-r)*(y-r)) / diameter);
```

Comment:
I think Math.sqrt(...) can be Math.hypot(x-r, y-r).
Is 0.70716... something to do with PI? If it is, using PI would be clearer...